### ISSUE TITLE
Add tracer to call device-plugin before

### ISSUE THREAD

**Issue by lengrongfu** [kubernetes/kubernetes#127051]
### What would you like to be added?

1. Add trace record in the call device-plugin before.
2. Transfer trace info(e.g pod info) to device-plugin server from gRPC header.

### Why is this needed?

Without modifying the API directly, I think the best way is to pass it to the device-plugin through Trace, and there are many similar requirements.

- https://github.com/kubernetes/kubernetes/issues/59109
- https://github.com/kubernetes/kubernetes/issues/107244

**Comment by lengrongfu** [comment]
@klueska What do you think?

**Comment by lengrongfu** [comment]
@SergeyKanzhelev What do you think?

**Comment by googs1025** [comment]
If I'm wrong, please correct me. In my opinion, this method of injecting information into headers is somewhat implicit. I don't think it's suitable for injecting details like pod information or container specifics in such an implicit manner. API changes might be more appropriate for passing such information, but I'm not quite sure why previous proposals haven't yielded many results. Additionally, I'm curious if there are specific use cases where a device plugin requires this additional information. Typically, from my understanding, scenarios like GPU topology scheduling involve a combination of device plugins and scheduler plugins, where scheduler plugins can access pod and node details to perform logical operations (e.g., filtering, scoring, etc.), while device plugins seem to focus solely on "allocating device resources." So, I wonder if there are any scenarios that necessitate extra information. Moreover, if additional information is needed, could we consider using DRA (Dynamic Resource Allocation)?

**Comment by lengrongfu** [comment]
Because we pod use device is scheduler choose, is not device-plugin choose.

**Comment by lengrongfu** [comment]
I tested this solution locally with very few changes. I just needed to package the pod information into ctx through metadata.
`pkg/kubelet/cm/devicemanager/manager.go`
```
md := metadata.Pairs(
      "pod_name", pod.Name,
      "pod_ns", pod.Namespace,
      "container_name", contName,
    )
metaCtx := metadata.NewOutgoingContext(context.Background(), md)
resp, err := eI.e.allocate(metaCtx, devs)
```

Then add the logic of reading pod information from metadata in `device-plugin`.

```
func (plugin *NvidiaDevicePlugin) Allocate(ctx context.Context, reqs *kubeletdevicepluginv1beta1.AllocateRequest) (*kubeletdevicepluginv1beta1.AllocateResponse, error) {
	md, ok := metadata.FromIncomingContext(ctx)
	if ok {
		podName := md["pod_name"]
		podNs := md["pod_ns"]
		containerName := md["container_name"]
		klog.Infof("Received pod_name: %v, pod_ns: %v, container_name: %v", podName, podNs, containerName)
	} else {
		klog.Warningf("Don't Received metadata info")
	}
}
```

**Comment by lengrongfu** [comment]
> I agree with @googs1025 , this will create a new API and contract on the device plugins over an unstable interface , IIRC there is another issue long time ago exactly about the same problem and discarded by the same reasons.
> 
> The project is investing in DRA to solve this problem, it will be bad we keep evolving device plugins instead of adding more efforts to graduate DRA.
> 
> For passing metadata to device plugins you can use the CDI interface, or just make your plugin connect to kubelet and the container runtime (via NRI) , I've presented last year how to do that with networking plugins https://docs.google.com/presentation/d/1pjDCtpdbCSWaqCbBYWgzTxAewOVbMf6rUS5SbjAJAe8/edit#slide=id.g2c2d098b0a2_0_427 so that part can be solved today without any modification

Thank you for your introduction. I will take a look at your design.

**Comment by SergeyKanzhelev** [comment]
If we only need this for tracing and logging we can consider supporting the Baggage (https://opentelemetry.io/docs/concepts/signals/baggage/) and blindly log everything Baggage contains.

It is NOT a good approach to extend the API using headers. We can extend the type `AllocateRequest` if we have enough justifications for this data be propagated.

**Comment by lengrongfu** [comment]
> If we only need this for tracing and logging we can consider supporting the Baggage (https://opentelemetry.io/docs/concepts/signals/baggage/) and blindly log everything Baggage contains.

vim pkg/kubelet/cm/devicemanager/manager.go
```
member1, _ := baggage.NewMember("pod_name", pod.Name)
member2, _ := baggage.NewMember("pod_ns", pod.Namespace)
member3, _ := baggage.NewMember("container_name", contName)
bag, _ := baggage.New(member1, member2, member3)
ctx := baggage.ContextWithBaggage(context.Background(), bag)
resp, err := eI.e.allocate(devs,ctx)
```
Is this code example what you meant? 

@SergeyKanzhelev

**Comment by SergeyKanzhelev** [comment]
> Is this code example what you meant?
> 
> @SergeyKanzhelev

Yes. And on other side - everything from the baggage can be logged. The key here is that we must not give any guarantees on what is inside the baggage and only recommend to use it for logging "as-is"
